core/src/main/java/org/elasticsearch/watcher/DirectoryWatcher.java

```diff
@@ -0,0 +1,257 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.watcher;
+
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.io.FileSystemUtils;
+import org.elasticsearch.common.logging.Loggers;
+
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+/**
+ * The file watcher checks directory and all its subdirectories for file changes and notifies its listeners accordingly
+ * http://docs.oracle.com/javase/tutorial/displayCode.html?code=http://docs.oracle.com/javase/tutorial/essential/io/examples/WatchDir.java
+ */
+public class DirectoryWatcher extends AbstractResourceWatcher<FileChangesListener> {
+
+    private Path file;
+    private final Map<WatchKey, Path> keys;
+
+    private WatchService watchService;
+    private WatchKey rootKey;
+
+    private static final Logger logger = Loggers.getLogger(DirectoryWatcher.class);
+
+    @SuppressWarnings("unchecked")
+    static <T> WatchEvent<T> cast(WatchEvent<?> event) {
+        return (WatchEvent<T>) event;
+    }
+
+    public DirectoryWatcher(Path file) throws IOException {
+        this.file = file;
+        this.watchService = FileSystems.getDefault().newWatchService();
+        this.keys = new HashMap<WatchKey, Path>();
+
+    }
+
+    @Override
+    protected void doInit() throws IOException {
+        rootKey = file.register(watchService, ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY);
+        if (!Files.isRegularFile(file)) {
+            registerAll(file, true);
+        }
+    }
+
+    @Override
+    protected void doCheckAndNotify() throws IOException {
+        while (true) {
+            // wait for key to be signalled
+            WatchKey key;
+            try {
+                key = watchService.take();
+            } catch (InterruptedException x) {
+                return;
+            }
+
+            Path dir = keys.get(key);
+            boolean isRoot = key.equals(rootKey);
+
+            if (dir == null && !isRoot) {
+                System.err.println("WatchKey not recognized!!");
```

Comment:
Hey you should run gradle check, it fails the test as we shouldn't be using System.err. Maybe we should their logger class.